### ISSUE TITLE
Add without-pam-u2f-nouserok

### ISSUE THREAD

**Issue by dlippold** [authselect/authselect#216]
With authselect version 1.2.1 (on an up-to-date Fedora 32) with used function "with-pam-u2f-2fa" (after executing "authselect select sssd with-pam-u2f-2fa without-nullok") it is still possible to login on the console without the Yubikey, in particular when the login manager or the screen lock is running. Therefore the "with-pam-u2f-2fa" is currently useless in respect of security.

How to reproduce (tested with Fedora LXDE 32):

1. Activate "with-pam-u2f-2fa" (e.g. execute "authselect select sssd with-pam-u2f-2fa without-nullok" by root)
2. Activate a Yubikey for a user (e.g. exectute "mkdir ~/.config/Yubico; pamu2fcfg > ~/.config/Yubico/u2f_keys" by the user and press the Yubikey button)
3. Reboot the system
4. Wait until the login manager appears
5. Press the key combination Strg-Alt-F2 to switch to a console
6. Login as the user only with the password without the Yubikey

Workaround:

Remove the option "nouserok" in the files "/etc/pam.d/*". But that should only be done when login by a Yubikey is already activated for root (otherwise root cannot login anymore).

**Comment by pbrezina** [comment]
Thank you for the report.

I understand the issue but I'm quite reluctant removing nouserok since it is fairly simple to get locked out of the system without this option - as you say, it will screw up `root` authentication if you forget to enable it fr root first or if you delete the file by accident.

I think the best thing to do would be to add new option to pam_u2f that would only allow root login without the key file. This way, you won't lock the root if the file is missing/deleted. Does this sound good to you?

And about the fact that you are able to login with password only from console when login manager is running, this sound like a bug in pam_u2f.

**Comment by dlippold** [comment]
> And about the fact that you are able to login with password only from console when login manager is running, this sound like a bug in pam_u2f.

I tested pam_u2f version 1.0.7-3 from Fedora 30 but that didn't changed the situation.

How can it be determined if the bug is in pam_u2f? What should be done in that case?

**Comment by dlippold** [comment]
> I think the best thing to do would be to add new option to pam_u2f that would only allow root login without the key file. This way, you won't lock the root if the file is missing/deleted. Does this sound good to you?

From my point of view it is most important to protect the root account by a second factor. Therefore there should be an option in authselect to avoid the adding of _nouserok_ (at least for root but better also for user accounts).

**Comment by pbrezina** [comment]
I'm going to answer the original comment since it actually seems important:

> From my point of view it is most important to protect the root account by a second factor. Therefore there should be an option in authselect to avoid the adding of nouserok (at least for root but better also for user accounts)

The truth is that you can't really enforce any authentication mechanism for root since as root you can do anything to the system. So if an attacker gets root access, he/she can of course change the pam configuration and remove the key file allowing easy access (though I really don't see why would anyone do this since they already have the access and this would only reveal their presence). So there is no extra security once they already have access.

What you can do for root is to make the key file immutable (`chattr +i file`) to prevent accidental deletion of the file but at the same time allowing the login without the file to prevent potential lock out.

But of course, enforcing it for unprivileged uses by removing `nouserok` make sense.

**Comment by pbrezina** [comment]
Let's try to bring attention to pam_u2f developers to get their opinion @a-dma

**Comment by dlippold** [comment]
> I'm going to answer the original comment since it actually seems important:
> 
> > From my point of view it is most important to protect the root account by a second factor. Therefore there should be an option in authselect to avoid the adding of nouserok (at least for root but better also for user accounts)
> 
> The truth is that you can't really enforce any authentication mechanism for root since as root you can do anything to the system. So if an attacker gets root access, he/she can of course change the pam configuration and remove the key file allowing easy access (though I really don't see why would anyone do this since they already have the access and this would only reveal their presence). So there is no extra security once they already have access.
> 
> What you can do for root is to make the key file immutable (`chattr +i file`) to prevent accidental deletion of the file but at the same time allowing the login without the file to prevent potential lock out.
> 
> But of course, enforcing it for unprivileged uses by removing `nouserok` make sense.

The original comment (which you cited) referred to the workaround I suggested for the time until the problem (login without Yubikey on console) is fixed. To protect (at least) the root account by a second factor there should be a possibility (the suggested option in authselect) to disable the login without Yubikey by preventing the adding of _nouserok_ . That should **prevent** that an attacked gets root access (until the problem is fixed, e.g. in pam_u2f).

But because the reference of my comment to the cited text of you was unclear I replaced it by a comment to the primary problem.

Independent of the complicated discussion I think we are agreed that:

1.  The bug should be fixed that it is currently possible to login on the console without a second factor when "with-pam-u2f-2fa" is used and **there is** a file "~/.config/Yubico/u2f_keys" 
2. When "with-pam-u2f-2fa" is used it should be obligatory for the users to have an entry in the authorization file (e.g. in "~/.config/Yubico/u2f_keys") but that should not be obligatory for root (i.e. if there is no entry for root root can login nevertheless).

We will see if these can be realized in pam_u2f near-term. If not we should discuss about the workarounds again.

**Comment by dlippold** [comment]
> Let's try to get attention of pam_u2f developers to get their opinion @a-dma

Did you already sent _a-dma_ a notice or should I add a bug report to https://github.com/Yubico/pam-u2f/issues ?

**Comment by a-dma** [comment]
Hi, I'm not familiar with authselect, but of course I'm willing to help. If you think that there is a bug in pam-u2f then please open an issue with a minimal example to reproduce it and I will have a look at it.

Keep in mind that pam-u2f is undergoing some major changes and the current git version is fairly different from the released one at this point.

**Comment by dlippold** [comment]
I have written a bug report for pam_u2f on https://github.com/Yubico/pam-u2f/issues/152

**Comment by dlippold** [comment]
I have written a RFE for pam_u2f regarding the second issue above on https://github.com/Yubico/pam-u2f/issues/154

**Comment by dlippold** [comment]
The first issue mentioned above is resolved by fixing a bug in SELinux in combination with an extension of the README file of pam_u2f. Every user who is already using pam_u2f has to update the security context of the configguration files. Details are described in the new README file of pam_u2f.

The second issue mentioned above will not result in a change of pam_u2f (see https://github.com/Yubico/pam-u2f/issues/154). Therefore I suggest to add the option `without-nouserok` to the profil _sssd_ of authselect.

**Comment by pbrezina** [comment]
Ok, added. Can you please test https://github.com/authselect/authselect/issues/216?

**Comment by pbrezina** [comment]
* `master`
  * db94f7c2902c87e6f637d29eccfd9a285fcb3783 - profiles: add without-pam-u2f-nouserok

**Comment by dlippold** [comment]
I have only just got around to do the test. And I found that the new option seems to work. But I would suggest to explicitly add to the description of the new option that root must have an u2f configuration. E.g. change the last sentence "Make sure ... privileges." to: "Make sure that root has an u2f configuration (e.g. there is the file `/root/.config/Yubico/u2f_keys`) and is able to log in. Otherwise you would lose root privileges."
